### PR TITLE
Redefine internal Kingdom proto messages to support v2alpha public API.

### PR DESCRIPTION
**Pull request by SanjayVas**
This is for #52.

This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)

### CODE REVIEW

**Review by efoxepstein — COMMENTED** [review]
> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZo9PX90Jtuc7__KCDO:b6a42at) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
> ```protobuf
> option java_multiple_files = true;
> 
> // Protocol configuration that need only be known by Duchies.
> ```

Does the Kingdom actually need an understanding of this if it's only needed by Duchies? (Can we treat it as bytes to avoid coupling this too tightly to the public API?)

Actually, more generally -- can we reduce what the Kingdom needs to understand at the data layer?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 3 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZoGRGp5vbEQ3Woj9aC:bwnsur7) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Does the Kingdom actually need an understanding of this if it's only needed by Duchies? (Can we treat it as bytes to avoid coupling this too tightly to the public API?)

Actually, more generally -- can we reduce what the Kingdom needs to understand at the data layer?
</blockquote></details>

My understanding is that this config must be the same for all Duchies participating in a given computation, but it's not something that MCs need to specify. The easiest way to do this is to have the Kingdom provide it (likely reading the values from some config file).

You're right that the actual values could be in an opaque bytes since the Kingdom doesn't do anything with the fields, but given that the Kingdom is the source for this it should also define the structure. IMO the only places where the Kingdom should persist opaque bytes is when it's acting as a pass-through, holding bytes from an external source.

**Review by efoxepstein — COMMENTED** [review]
> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZoJpAi85NXEiDM_AhH:b-r9948z) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

My understanding is that this config must be the same for all Duchies participating in a given computation, but it's not something that MCs need to specify. The easiest way to do this is to have the Kingdom provide it (likely reading the values from some config file).

You're right that the actual values could be in an opaque bytes since the Kingdom doesn't do anything with the fields, but given that the Kingdom is the source for this it should also define the structure. IMO the only places where the Kingdom should persist opaque bytes is when it's acting as a pass-through, holding bytes from an external source.
</blockquote></details>

Would it make sense for the Kingdom to understand the structure at the public API level but not at the data layer then?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZoKSmZ1XbbdYWnJ11v:b-do1jz8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Would it make sense for the Kingdom to understand the structure at the public API level but not at the data layer then?
</blockquote></details>

It needs to understand it sufficiently at the persistence layer to make conversions between API versions, since the same persistence may need to support more than one API version (assuming they are sufficiently compatible but not necessarily binary-compatible).

I suppose we could theoretically persist one API version and convert to others, but that means you need to either keep old API definitions around for reading old rows or do a data migration. (This particular message will most likely be in a config file so migration should be easy, but I don't want to make that assumption in the proto definition).

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @SanjayVas, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZoMUEK9rUXYvJQpQuM:b-55f760) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
Don't we have this problem with all sorts of serialized things like this?

> ```
>  bytes el_gamal_public_key = 1;
> ```

and 

```
// Serialized `MeasurementSpec` from public API.
bytes measurement_spec = 2;
```

I guess one philosophy is enforce a consistent API version for all parties per Measurement and then hope we never need to surface these details for old measurements in UIs.

I don't particularly care what we do here with DuchyProtocolConfig but we should have some philosophy as we're likely to run into this a few more times.

**Review by kungfucraig — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZoPRvzEp7oTovdhVG6:bg4v3x8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Don't we have this problem with all sorts of serialized things like this?

> ```
>  bytes el_gamal_public_key = 1;
> ```

and 

```
// Serialized `MeasurementSpec` from public API.
bytes measurement_spec = 2;
```

I guess one philosophy is enforce a consistent API version for all parties per Measurement and then hope we never need to surface these details for old measurements in UIs.

I don't particularly care what we do here with DuchyProtocolConfig but we should have some philosophy as we're likely to run into this a few more times.
</blockquote></details>

All other things begin equal, I think erring on the side of explicitly storing things is better. At the very least it helps debugging.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @wangyaopw)

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @kungfucraig, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 24 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZo9PX90Jtuc7__KCDN:-MZonq4C3RPVHF5-3ObW:b394iz2) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/310213271eff6d1e54a5703f10358c495aa63df9/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L24)):*
> Don't we have this problem with all sorts of serialized things like this?

> I guess one philosophy is enforce a consistent API version for all parties per Measurement and then hope we never need to surface these details for old measurements in UIs.

Yes, in some cases the Kingdom is serving as a pass-through for pre-serialized values where the actual bytes are significant (regardless if they deserialize into an equivalent message). In those cases, we can't do much beyond checking or surfacing the API version of the serialized messages to prevent clients from deserializing incorrectly. This is why this PR includes fields for persisting API version. (We likely need to surface these in the API so that old entries can be accessed even on newer API versions).

IMO we should only persist API messages when we don't have a choice.

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 1 unresolved discussion (waiting on @kungfucraig and @wangyaopw)

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 2 unresolved discussions (waiting on @kungfucraig and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZsIAB24RRN3bGilZKY:-MZsIAB24RRN3bGilZKZ:b-kkv9k4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/ab803332ffd8ab68e4d8fedc0d17cf12a6d73461/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L27)):*
> ```
>     // The maximum frequency to reveal in the histogram.
>     int32 maximum_frequency = 1;
> ```
@wangyaopw I assume this is something that MCs shouldn't be free to decide, so they can't specify it in the measurement spec. Should this perhaps be in the public protocol config so that MCs can know the max frequency after the measurement is created?

**Review by wangyaopw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 2 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto, line 27 at r4](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZsIAB24RRN3bGilZKY:-MZsOv5y22_b-89E69AV:bqnbmon) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/ab803332ffd8ab68e4d8fedc0d17cf12a6d73461/src/main/proto/wfa/measurement/internal/kingdom/duchy_protocol_config.proto#L27)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
>     // The maximum frequency to reveal in the histogram.
>     int32 maximum_frequency = 1;
> ```
@wangyaopw I assume this is something that MCs shouldn't be free to decide, so they can't specify it in the measurement spec. Should this perhaps be in the public protocol config so that MCs can know the max frequency after the measurement is created?
</blockquote></details>

Technically, it is OK for the MC to specify the maximum_frequency. But since it impacts the computation cost, we won't allow them to wildly choose a value. If we set a cap on this value, e.g., 15, then I guess most MCs would just use 15 for it. so it seems not necessary to make it configureable by the MC at all.

The last bucket in the final report would be for "K+", so MC would know what max frequency is used.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 3 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto, line 43 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZuPyyG9tavFhUrT2tb:-MZuPyyG9tavFhUrT2tc:b9cc23c) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/fcca129dd1c33be0d13ede156b655e34642b0a2d/src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto#L43)):*
> ```protobuf
>     REQUISITION_PARAMS_SET = 2;
> 
>     // This `ComputationParticipant` is ready to participate.
> ```

Is this the terminal state? Are there any others we need like "REFUSED" or something?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 8 files reviewed, 1 unresolved discussion (waiting on @efoxepstein)

___
*[src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto, line 43 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZuPyyG9tavFhUrT2tb:-MZyWqyT7H6XnfgoaRT9:b50p4rn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/fcca129dd1c33be0d13ede156b655e34642b0a2d/src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto#L43)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Is this the terminal state? Are there any others we need like "REFUSED" or something?
</blockquote></details>

Yes to both. This is just reflecting the states from the system API, so also updating in #44.

Let me know if you think we need to capture refusal info like we do for Requisitions. I assume the only real case for refusal would be signature verification failure, and most other errors would come from computation stage attempts.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto, line 43 at r5](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZuPyyG9tavFhUrT2tb:-MZy_wMLEhAR72zqwyXJ:bdjwouw) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/fcca129dd1c33be0d13ede156b655e34642b0a2d/src/main/proto/wfa/measurement/internal/kingdom/computation_participant.proto#L43)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Yes to both. This is just reflecting the states from the system API, so also updating in #44.

Let me know if you think we need to capture refusal info like we do for Requisitions. I assume the only real case for refusal would be signature verification failure, and most other errors would come from computation stage attempts.
</blockquote></details>

This seems fine for now. We can always add a refusal justification later if needed.

File: src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+option java_package = "org.wfanet.measurement.internal.kingdom";
+option java_multiple_files = true;
+
+// Parameters for differential privacy (DP).
+//
+// For detail, refer to "Dwork, C. and Roth, A., 2014. The algorithmic
+// foundations of differential privacy. Foundations and Trends in Theoretical
+// Computer Science, 9(3-4), pp.211-407."
+message DifferentialPrivacyParams {
```
Comment (wangyaopw) [inline src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto:27]: I am wondering if we should define such proto in the common repo. They are unlikely to change in the future.

I already created one in any-sketch and is referring to it inside the duchy. We could move it to the common repo and just use it everywhere.. 
 https://github.com/world-federation-of-advertisers/any-sketch/blob/bd54716d9c76664df363c0b16337c09443074b09/src/main/proto/wfa/common/noise_parameters.proto

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 9 files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @kungfucraig, and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_1dFR-22dfGCQ2d3ns:b-ewmjmk) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

I am wondering if we should define such proto in the common repo. They are unlikely to change in the future.

I already created one in any-sketch and is referring to it inside the duchy. We could move it to the common repo and just use it everywhere.. 
 https://github.com/world-federation-of-advertisers/any-sketch/blob/bd54716d9c76664df363c0b16337c09443074b09/src/main/proto/wfa/common/noise_parameters.proto
</blockquote></details>

It's a bad idea to depend on external proto messages in either persistence or APIs. The exceptions is for stable well-known messages, such as `google.protobuf.Timestamp`.

This is why we have some messages that are currently duplicated in the public API and internal persistence, since they may evolve somewhat independently (e.g. in the case where an API proto change can be implemented without a data migration).

I don't know if I have sufficiently high confidence in any of our messages not changing, so I'd rather that we can isolate them even at the cost of duplication.

**Comment by SanjayVas** [comment]
Added `MeasurementLogEntry`, intended to replace `ReportLogEntry`. PTAL.

**Review by wangyaopw — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 7 of 9 files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @kungfucraig, and @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_1h3lKBb1e9i0xDvD-:b-2phyhb) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

It's a bad idea to depend on external proto messages in either persistence or APIs. The exceptions is for stable well-known messages, such as `google.protobuf.Timestamp`.

This is why we have some messages that are currently duplicated in the public API and internal persistence, since they may evolve somewhat independently (e.g. in the case where an API proto change can be implemented without a data migration).

I don't know if I have sufficiently high confidence in any of our messages not changing, so I'd rather that we can isolate them even at the cost of duplication.
</blockquote></details>

These math protos are very unlikely to change no matter what system/protocol we are building.

**Review by kungfucraig — APPROVED** [review]

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 6 of 19 files reviewed, 1 unresolved discussion (waiting on @efoxepstein)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_1pqsxBqmXuo52PyEP:b-hw141) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, wangyaopw (Yao Wang) wrote…</i></summary><blockquote>

These math protos are very unlikely to change no matter what system/protocol we are building.
</blockquote></details>

@efoxepstein Seeking second opinion. Do you think this would be stable enough to consider it a common type? This implies that we would put in in a common `type` package and share it among internal and API services as described in https://google.aip.dev/215.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 1 of 2 files at r7, 12 of 12 files at r8.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 1 unresolved discussion (waiting on @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_1tZYl5KqhrIwrqk9y:bzcwvty) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

@efoxepstein Seeking second opinion. Do you think this would be stable enough to consider it a common type? This implies that we would put in in a common `type` package and share it among internal and API services as described in https://google.aip.dev/215.
</blockquote></details>

I think the risk here is pretty low.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 17 of 19 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_1wPHK3IWmgFRW8Ufb:b-ejkvak) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

I think the risk here is pretty low.
</blockquote></details>

So I believe the following repos would depend on this common type: any-sketch, cmm, and cmm-api. cmm depends on both cmm-api and any-sketch.

Any thoughts on where we should put this? Should it be a new repo for common protobuf types? Package `halo.type`?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 17 of 19 files reviewed, 1 unresolved discussion (waiting on @efoxepstein and @wangyaopw)

___
*[src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto, line 27 at r7](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-MZZgXrBNd9GPjkJCR0Q-r7-27:-M_2_rqqDzDNyZCrhSg5:b80u00e) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/a44ff5b99ffc18648a336b673ff0b2fc5c95794b/src/main/proto/wfa/measurement/internal/kingdom/differential_privacy.proto#L27)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

So I believe the following repos would depend on this common type: any-sketch, cmm, and cmm-api. cmm depends on both cmm-api and any-sketch.

Any thoughts on where we should put this? Should it be a new repo for common protobuf types? Package `halo.type`?
</blockquote></details>

From offline discussion, we've agreed to keep these separate out of caution since proto converters aren't too difficult to write/maintain.

**Review by wangyaopw — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 17 of 19 files reviewed, all discussions resolved (waiting on @efoxepstein)

**Review by efoxepstein — APPROVED** [review]
> Reviewed 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 19 of 25 files reviewed, 3 unresolved discussions (waiting on @corbantek, @oliver-amzn, @stevenwarejones, and @zachcwc)

___
*[src/main/proto/wfa/measurement/internal/kingdom/requisition.proto, line 57 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-M_6-3H842juDH1xFxEf:-M_7-PTf2zov6HdWfy6O:b-293ryi) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/044c22a60b008b549ccbfe19b9bbbd0c2bc225fb/src/main/proto/wfa/measurement/internal/kingdom/requisition.proto#L57)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

what would a duplicate hash/salt/requisition be considered?
</blockquote></details>

I think that fits under this case, since every measurement should have a unique key.

Note that this is currently just a copy of the enum from the v2alpha public API.
___
*[src/main/proto/wfa/measurement/internal/kingdom/requisition.proto, line 93 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-M_6-bJB9LPVrLha9OSm:-M_7-jns-qRMm8_WQ5zp:biz0k8i) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/044c22a60b008b549ccbfe19b9bbbd0c2bc225fb/src/main/proto/wfa/measurement/internal/kingdom/requisition.proto#L93)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

is this the data provider's encryption key?
</blockquote></details>

Yes. It's to persist that data from the `Requisition` API resource.
___
*[src/main/proto/wfa/measurement/internal/kingdom/requisition.proto, line 94 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-M_6-wfl0EGD4hKzs0vh:-M_7003J4kH3YMM1OBm_:bw35fc8) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/044c22a60b008b549ccbfe19b9bbbd0c2bc225fb/src/main/proto/wfa/measurement/internal/kingdom/requisition.proto#L94)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

Is this the public signing key for this particular requisition signed with the root certificate?
</blockquote></details>

This is the signature of the above public key field, which can be verified using the certificate indicated by the `external_data_provider_certificate_id` field.

I can add additional documentation for this, but it's just for persisting the equivalent data from the public API.

**Review by stevenwarejones — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 18 of 25 files reviewed, 3 unresolved discussions (waiting on @corbantek, @efoxepstein, @oliver-amzn, @SanjayVas, @stevenwarejones, and @zachcwc)

___
*[src/main/proto/wfa/measurement/internal/kingdom/requisition.proto, line 93 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-M_6-bJB9LPVrLha9OSm:-M_77hiYBudvdbA1T12j:b-f34xwy) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/044c22a60b008b549ccbfe19b9bbbd0c2bc225fb/src/main/proto/wfa/measurement/internal/kingdom/requisition.proto#L93)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Yes. It's to persist that data from the `Requisition` API resource.
</blockquote></details>

what do you think of naming it to data_provider_public_encryption_key? So, its not confused with a signing key.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: 17 of 25 files reviewed, 3 unresolved discussions (waiting on @corbantek, @efoxepstein, @oliver-amzn, @stevenwarejones, and @zachcwc)

___
*[src/main/proto/wfa/measurement/internal/kingdom/requisition.proto, line 93 at r10](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39#-M_6-bJB9LPVrLha9OSm:-M_7AAzX82EWFUKjuNxu:benspis) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/044c22a60b008b549ccbfe19b9bbbd0c2bc225fb/src/main/proto/wfa/measurement/internal/kingdom/requisition.proto#L93)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

what do you think of naming it to data_provider_public_encryption_key? So, its not confused with a signing key.
</blockquote></details>

I updated the comment to indicate that this is a serialized `EncryptionPublicKey` message, which not only makes the purpose clear but also indicates that it includes the key type.

As an aside, signing keys aren't actually ever persisted/exposed by the system on their own. Those are always part of an X.509 certificate, which would be in a field that has `certificate` somewhere in the name. Therefore, any public key field in the system is an encryption key.

**Review by efoxepstein — APPROVED** [review]
> Reviewed 6 of 6 files at r10, 2 of 2 files at r11.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/39)* status: all files reviewed, 3 unresolved discussions (waiting on @corbantek, @oliver-amzn, @stevenwarejones, and @zachcwc)

**Pull request merged**